client/finality-grandpa/src/lib.rs

```diff
@@ -1089,32 +1095,44 @@ pub fn setup_disabled_grandpa<Block: BlockT, Client, N>(
 /// Returns the key pair of the node that is being used in the current voter set or `None`.
 fn is_voter(
 	voters: &Arc<VoterSet<AuthorityId>>,
-	keystore: &Option<KeyStorePtr>,
-) -> Option<AuthorityPair> {
+	keystore: &Option<BareCryptoStorePtr>,
+) -> Option<AuthorityId> {
 	match keystore {
 		Some(keystore) => voters
 			.iter()
-			.find_map(|(p, _)| keystore.read().key_pair::<AuthorityPair>(&p).ok()),
+			.find_map(|(p, _)| {
+				let can_vote = keystore.read()
+					.has_keys(&[(p.to_raw_vec(), AuthorityId::ID)]);
+				if can_vote {
+					Some(p.clone())
+				} else {
+					None
+				}
+			}),
 		None => None,
 	}
 }
 
 /// Returns the authority id of this node, if available.
 fn authority_id<'a, I>(
 	authorities: &mut I,
-	keystore: &Option<KeyStorePtr>,
+	keystore: &Option<BareCryptoStorePtr>,
 ) -> Option<AuthorityId> where
 	I: Iterator<Item = &'a AuthorityId>,
 {
 	match keystore {
 		Some(keystore) => {
 			authorities
 				.find_map(|p| {
-					keystore.read().key_pair::<AuthorityPair>(&p)
-						.ok()
-						.map(|ap| ap.public())
+					let can_vote = keystore.read()
+						.has_keys(&[(p.to_raw_vec(), AuthorityId::ID)]);
+					if can_vote {
+						Some(p.clone())
+					} else {
+						None
+					}
 				})
-		}
+		},
```

Comment:
```
.find(|p| keystore.read().has_keys(&[(p.to_raw_vec(), AuthorityId::ID)]).cloned()
```

Reply:
`has_keys` returns a `bool` in this case while the function needs to return the public key.

Reply:
`find`, unlike `find_map`, takes a predicate and returns the first element that matches this predicate. So in this case the example from @bkchr should work since we don't need to map the input.